### PR TITLE
added Normal inverse gamma distribution

### PR DESCRIPTION
**Pull request by parulsethi**
This PR implements the Normal inverse gamma distribution which is a conjugate prior for Normal distribution with unknown mean and variance. I have added it under _multivariate.py as suggested by @josef-pkt in [#6528](https://github.com/scipy/scipy/issues/6528) with few basic methods. As it's a bivariate distribution i've addressed the unknown mean's variate as `x` and unknown variance's variate as `sig2` separately instead of using a 2D variate for now.
Please suggest any changes required for improving this class.
I will write the documentation and unit tests as soon as the methods are finalised for this distribution.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -50,6 +50,8 @@ _configtest.c
 
 # Python files #
 ################
+# dev directory
+scipy-dev
```
Comment (rgommers) [inline .gitignore]: this shouldn't be here, that's specific to your install (and looks wrong)

**Comment by rgommers** [comment]
Thanks @parulsethi. 

From a first glance this looks good. It will need tests of course, those are missing still. There's two issues in the docstring: https://travis-ci.org/scipy/scipy/jobs/171665860#L3162

File: scipy/stats/_multivariate.py [hunk 2]
```diff
@@ -3521,3 +3522,353 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
...
+
+    Methods
+    -------
+    ``pdf(x, sig2, loc=0, variance_scale=1, shape=1, scale=1)``
+        Probability density function.
+    ``logpdf(x, sig2, loc=0, variance_scale=1, shape=1, scale=1)``
+        Log of the probability density function.
+    ``rvs(loc=0, variance_scale=1, shape=1, scale=1, size=1, random_state=None)``
+        Draw random samples from normal and inverse gamma distribution.
+    ``mean(loc=0, variance_scale=1, shape=1, scale=1)``
```
Comment (rgommers) [inline scipy/stats/_multivariate.py]: The ``mean`` and ``mode`` methods don't seem to do much. Other distributions do have them, so maybe worth keeping just for consistency. The descriptions should be changed though, there are no random variates involved here.

File: scipy/stats/_multivariate.py [hunk 3]
```diff
@@ -3521,3 +3522,353 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
...
+        Mean of the random variates.
+    ``mode(loc=0, variance_scale=1, shape=1, scale=1)``
+        Mode of the random variates.
+
+    Parameters
+    ----------
+    x : array
+            One-dimensional array.
```
Comment (rgommers) [inline scipy/stats/_multivariate.py]: indentation is off here, should be 4 spaces

File: scipy/stats/_multivariate.py [hunk 4]
```diff
@@ -3521,3 +3522,353 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
...
+
+    Examples
+    --------
+    >>> import matplotlib.pyplot as plt
+    >>> from mpl_toolkits.mplot3d import axes3d
+    >>> from scipy.stats import normal_invgamma
+    >>> x = np.linspace(0, 5, 10, endpoint=False)
+    >>> sig2 = np.linspace(5, 10, 10, endpoint=False)
+    >>> z = normal_invgamma.pdf(x, sig2, loc=2,
+                                    variance_scale=3, shape=5, scale=2); z
```
Comment (rgommers) [inline scipy/stats/_multivariate.py]: never use ``;``, just do ``>>> z`` on the next line

File: scipy/stats/_multivariate.py [hunk 5]
```diff
@@ -3521,3 +3522,353 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
...
+    array([  5.15655226e-06,   3.07181064e-06,   1.87273568e-06,
+         1.16664399e-06,   7.41424234e-07,   4.79916138e-07,
+         3.15923431e-07,   2.11211796e-07,   1.43229349e-07,
+         9.84088858e-08])
```
Comment (rgommers) [inline scipy/stats/_multivariate.py]: probably worth getting a larger number of variates. I haven't checked the plot, but I suspect it'll not say anything interesting about the shape of this distribution.

File: scipy/stats/_multivariate.py [hunk 6]
```diff
@@ -3521,3 +3522,353 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
...
+        """
+        Create a frozen normal inverse gamma distribution.
+
+        See `normal_invgamma_frozen` for more information.
+
+        """
+        return normal_invgamma_frozen(loc, variance_scale, shape, scale,
+                                    seed=seed)
+
+    def _check_parameters(self, loc, variance_scale, shape, scale):
```
Comment (rgommers) [inline scipy/stats/_multivariate.py]: Why should all parameters be scalars? Other distributions allow for array-like parameters, and do broadcasting if that makes sense.

File: scipy/stats/_multivariate.py [hunk 7]
```diff
@@ -3521,3 +3522,353 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
...
+        pdf : ndarray
+            Log of the probability density function evaluated at `(x, sig2)`.
+
+        """
+        loc, variance_scale, shape, scale = self._check_parameters(
+                                                    loc, variance_scale, shape, scale)
+        x, sig2 = self._check_input(x, sig2)
+
+        Zinv = shape * np.log(scale) - gammaln(shape) - 0.5 * (np.log(variance_scale) + _LOG_2PI)
+        out = Zinv - 0.5 * np.log(sig2) - (shape + 1.) * np.log(sig2) -\
```
Comment (rgommers) [inline scipy/stats/_multivariate.py]: style: no ``\`` please (use brackets instead)

**Comment by rgommers** [comment]
Looks like there is a closed-form solution for ``cdf``, so that might make sense to add: https://en.wikipedia.org/wiki/Normal-inverse-gamma_distribution#Cumulative_distribution_function

**Comment by rgommers** [comment]
sig2, loc=0, variance_scale=1, shape=1, scale=1

That's a lot of parameters that ``invwishart`` doesn't have, and ``normal_invgamma`` is supposed to be a special case of ``invwishart``. Any thoughts on that?

**Comment by rgommers** [comment]
R has a ``norminvgamma`` distribution in the PIGShift package, may be good to check that results are consistent with that.

**Comment by rgommers** [comment]
@josef-pkt had some suggestions for additional methods here: https://github.com/scipy/scipy/issues/6528#issuecomment-252428345

File: scipy/stats/_multivariate.py (hunk 6 above)
Comment (parulsethi) [inline scipy/stats/_multivariate.py]: Yes, their shape should be checked from the dimension of `x`, sorry, i'll correct this.

**Comment by parulsethi** [comment]
Thanks for the R suggestion, they have a difference of one parameter ie. sig2, and the one i followed for pdf is corresponding to this package [JuliaStats](https://github.com/JuliaStats/ConjugatePriors.jl/blob/master/src/normalinversegamma.jl), and took sig2 separately. Please suggest for what convention to follow here.  
Inverse wishart is a prior just over the co-variance matrix, whereas this one is a prior over both mean and variance, hence takes two more hyperparameters.

**Comment by ev-br** [comment]
`pdf` should be able to accept the output of `rvs`, but

```
In [6]: r = stats.normal_invgamma.rvs(size=4)

In [7]: stats.normal_invgamma.pdf(r)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-cd28b06ba616> in <module>()
----> 1 stats.normal_invgamma.pdf(r)

TypeError: pdf() takes at least 3 arguments (2 given)
```

File: scipy/stats/_multivariate.py [hunk 8]
```diff
@@ -3521,3 +3522,353 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
...
+        %(_nig_doc_callparams_note)s
+
+        """
+        loc, variance_scale, shape, scale = self._check_parameters(
+                                                loc, variance_scale, shape, scale)
+        random_state = self._get_random_state(random_state)
+
+        sig2_rv = 1/random_state.gamma(shape, scale, size)
+        x_rv = random_state.normal(loc, np.sqrt(sig2_rv * variance_scale), size)
+        return np.array(list(zip(x_rv, sig2_rv)))
```
Comment (ev-br) [inline scipy/stats/_multivariate.py]: `np.c_[x_rv, sig2_rv]`

**Comment by parulsethi** [comment]
Oh, actually pdf takes 2 inputs separately which rvs is returning in zipped list, i'll make pdf to accept it accordingly.
But, i'm stiil doubtful over the sig2 parameter as how this [R package(dmvnorminvgamma)](https://github.com/cran/PIGShift/blob/01025e9ff7eb8d2d9a4bcb1b9125a0ceeb203d70/R/brownian_motion_package.r) suggested is going without it. It is the support for inverse gamma part of distribution. I'll appreciate some help with that.

File: scipy/stats/_multivariate.py [hunk 9]
```diff
@@ -3521,3 +3522,353 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
...
+    .. math::
+
+        f(x,\sigma^2) = \frac {1} {\sigma\sqrt{2\pi\nu} } \, \frac{\beta^\alpha}
+                        {\Gamma(\alpha)} \, \left( \frac{1}{\sigma^2} \right)^
+                        {\alpha + 1}\exp \left( \frac { -\beta}{\sigma^2} -
+                        \frac{(x - \mu)^2} {2\sigma^2\nu}  \right),
+
+    where :math:`\mu` is the loc, :math:`\nu` the variance scale,
+    :math:`\alpha` the shape, :math:`\beta` the scale and :math:`x` and
+    `\sigma^2` takes values.
```
Comment (ev-br) [inline scipy/stats/_multivariate.py]: > :math:`x` and   `\sigma^2` takes values.

Not sure what "takes values" means --- did you mean to finish the sentence?

Comment (ev-br) [inline scipy/stats/_multivariate.py]: also minor nit:  need `:math:` in front of `sigma^2` for consistency.

Comment (ev-br) [inline scipy/stats/_multivariate.py]: We typically encourage meaningful names. However here, I'm not sure "shape", "loc" and "variance_scale"  are worth it (besides, "shape" is used for general shape parameters for univariate distributions). Maybe rename them back to `alpha`, `beta` and `nu` in function signatures and documentation.

File: scipy/stats/_multivariate.py [hunk 10]
```diff
@@ -3521,3 +3522,353 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
...
+    >>> sig2 = np.linspace(5, 10, 10, endpoint=False)
+    >>> z = normal_invgamma.pdf(x, sig2, loc=2,
+                                    variance_scale=3, shape=5, scale=2); z
+    array([  5.15655226e-06,   3.07181064e-06,   1.87273568e-06,
+         1.16664399e-06,   7.41424234e-07,   4.79916138e-07,
+         3.15923431e-07,   2.11211796e-07,   1.43229349e-07,
+         9.84088858e-08])
+    >>> fig = plt.figure()
+    >>> ax = fig.add_subplot(111, projection='3d')
+    >>> ax.plot_surface(x, sig2, z)
```
Comment (ev-br) [inline scipy/stats/_multivariate.py]: Needs a `>>> plt.show()` to render the plot, see https://1352-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.stats.normal_invgamma.html#scipy.stats.normal_invgamma

**Comment by ev-br** [comment]
> Oh, actually pdf takes 2 inputs separately which rvs is returning in zipped list, i'll make pdf to accept it accordingly.
But, i'm stiil doubtful over the sig2 parameter as how this R package(dmvnorminvgamma) suggested is going without it. It is the support for inverse gamma part of distribution. I'll appreciate some help with that.

and

> Thanks for the R suggestion, they have a difference of one parameter ie. sig2, and the one i followed for pdf is corresponding to this package JuliaStats, and took sig2 separately. Please suggest for what convention to follow here.

So the issue is that Julia stats decided to have `x` and `sigma` separate, while R's package seems to pack them into a single variable? 

From a quick glance at https://cran.r-project.org/web/packages/PIGShift/PIGShift.pdf it seems to me that they have T as the "the variance-covariance matrix", which corresponds to `V` in Wikipedia.

BTW, better be careful with looking at the PIGShift code (and R code in general): it's GPL, and we cannot base scipy code on anything GPL.

What you can do though, you can read their docs, run R code and check the results. It would actually be very good to establish the correspondence and add tests comparing the values and listing the exact R code, like so: https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_multivariate.py#L943

Ditto for Julia code if you know how to run it :-).

As to the signatures of `pdf` and `rvs`: I'm fairly certain that trying to pack `x` and `sigma` into a 2D array would be *very* confusing. Better separate them into `x, sigma` or `x, sigma2`.

File: scipy/stats/_multivariate.py [hunk 11]
```diff
@@ -3521,3 +3522,353 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
         return m
 
 random_correlation = random_correlation_gen()
+
+
+_nig_doc_default_callparams = """\
+loc : scalar, optional
+    Mean of the distribution (default zero)
+variance_scale : positive scalar, optional
+    Scale on normal distribution prior (default one)
```
Comment (ev-br) [inline scipy/stats/_multivariate.py]: nit: default *is* one.

**Comment by ev-br** [comment]
Oh, and,
> I will write ... unit tests as soon as the methods are finalised for this distribution.

Please do write them!

**Comment by Naereen** [comment]
Hi!
Any update on this?

**Comment by rgommers** [comment]
> Hi! Any update on this?

Hi @Naereen, this PR seems to have stalled. Would be great if @parulsethi would pick it up and address the review comments, but not sure that's going to happen after more than half a year. If you're interested, you could take over this PR and finish it.

**Comment by Naereen** [comment]
Hi @rgommers.
I would love to but honestly won't have the time, sorry.

**Comment by parulsethi** [comment]
Sorry, I won't be able to take this up in near future at least. Any takers for this PR are welcome. 
I'll try to revisit after around 2-3 months and address review if no one takes it up till that.

**Comment by chrisb83** [comment]
Hi,

I would like to take up this topic, but might need some help since I am not so familiar with github yet.

I addressed quite a few of the review comments above and managed to push the commit to my forked repo https://github.com/chrisb83/scipy.git, branch norminvgamma. I changed the name to norminvgamma similar to #2648 and as in R package PIGshift.

I would also suggest to simplify the parametrisation: 
in my view, one only needs two parameters a and b to characterize the inverse Gamma "sigma2". Using the notation of Wikipedia (https://en.wikipedia.org/wiki/Normal-inverse-gamma_distribution), I would argue that one can take mu=0 and Lambda = 1 by default. If 

`(X, sigma2) norminvgamma(a,b,0,1), `

then 

`(mu + X/sqrt(Lambda), sigma2) norminvgamma(a, b, mu, Lambda). 

`Dropping mu and Lambda in the implementation makes the code much simpler.

R package PIGshift also uses this form: `dnorminvgamma(x, alpha, beta, log = FALSE).`

Feedback welcome!

Question: how can I commit against this pull request?

**Comment by rgommers** [comment]
> Question: how can I commit against this pull request?

This is only possible if @parulsethi gives you commit rights. The alternative is just to take over this branch and open a new pull request yourself. This PR has been stalled for quite a while, so unless you hear from @parulsethi within a day or so that would be fine to do.

**Comment by rgommers** [comment]
> similar to #2648 and as in R package PIGshift.

Just in case: `PIGshift` is GPL licensed, so please keep in mind that you cannot read its source code when you're working on an implementation for SciPy. Only reading the documentation for that package is fine though.

**Comment by Naereen** [comment]
> PIGshift is GPL licensed, so please keep in mind that you cannot read its source code when you're working on an implementation for SciPy

I don't think the GPL license is as strict as this, you cannot copy-paste the code, but reading it to understand the underlying algorithms cannot be forbidden...

**Comment by rgommers** [comment]
> I don't think the GPL license is as strict as this, you cannot copy-paste the code, but reading it to understand the underlying algorithms cannot be forbidden...

Yes it definitely is. There are lots of gray areas around GPL, but this is not one of them.

**Comment by chrisb83** [comment]
Hi all,

I worked a bit more on this topic and I think I have a version now that I pushed to my repository. The more I looked at this topic, the more I began to wonder how useful this distribution is. 

If `(X, sigma2)` is normalinvgamma, then sigma2 is simply inverse Gamma(a, b) and `X = loc + scale * sqrt(sigma2) * Z` is some essentially a Student t-distribution (because `sigma2 = 1/Gamma-distribution(a,1/b) = 1/Chisquare-distribution` (with certain parameters, scaled by constant), and Z/sqrt(chisquare) is Student-t). So the marginal distributions are already known and available in scipy.

Main application seems to be Baysian statistics (distribution used as a prior for unknown mean and variance of normal distribution). I only have basic knowledge of Baysian statistics and do not know what features would be desirable from this perspective. I guess pdf could be useful to get the posterior... 

If it makes sense to add this distribution with main methods being PDF and rvs, I can create a pull request.

Further comments:
- CDF stated on wikipedia does not seem correct (if sigma2 -> inf, the CDF goes to zero if I am not mistaken), see https://en.wikipedia.org/wiki/Normal-inverse-gamma_distribution. I haven't seen a closed form anywhere else
- dnorminvgamma in Rpackage PIGshift seems to be something else, i think it is a Student-t distribution, which would be the marginal distribution of X if (X, sigma2) is norminvgamma (with the right parameters)
- not sure what input should be allowed for the parameters; current implementation is probably not rigid enough and could lead easily to issues when trying to broadcast. maybe better to only allow scalar values.

**Comment by kleinschmidt** [comment]
The formula for mode is (I think) for the modes of the marginal distributions, not the joint distribution.  See JuliaStats/ConjugatePriors.jl#21

**Comment by tdgunes** [comment]
Hi there, are there any updates on this PR? 

_I'm planning to use this as a prior to a normal distribution in a probabilistic graphical model. It'd be great to have this in SciPy._

**Comment by chrisb83** [comment]
@tdgunes Hi, what functionality are you looking for? Generation of rvs? I think no one is working on this PR at the moment, but you could use the code already committed.

**Comment by kleinschmidt** [comment]
Here's a derivation of the mode (for normal-inverse chi sq, with the parametrization for normal inverse gamma at the end): https://gist.github.com/kleinschmidt/0466d46aed1736b47a01e5b975e07185

**Comment by vyasr** [comment]
I believe that this PR was resolved in pull request #8171 and can be closed now.

File: scipy/stats/_multivariate.py [hunk 12]
```diff
@@ -3521,3 +3522,353 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
...
+            Log of the probability density function evaluated at `(x, sig2)`.
+
+        """
+        loc, variance_scale, shape, scale = self._check_parameters(
+                                                    loc, variance_scale, shape, scale)
+        x, sig2 = self._check_input(x, sig2)
+
+        Zinv = shape * np.log(scale) - gammaln(shape) - 0.5 * (np.log(variance_scale) + _LOG_2PI)
+        out = Zinv - 0.5 * np.log(sig2) - (shape + 1.) * np.log(sig2) -\
+                scale/sig2 - 0.5/(sig2 * variance_scale) * (x-loc)**2
```
Comment (AndrewLim1990) [inline scipy/stats/_multivariate.py]: Is there a mistake here? I think it should be 
`scale/sig2 - 0.5/sig2 * variance_scale * (x-loc)**2`
(Removed brackets)

**Comment by Cartman0** [comment]
i want normal-inv-gamma.pdf, is there already ?

**Comment by mdhaber** [comment]
I'll go ahead and close this. The original issue https://github.com/scipy/scipy/issues/6528 tracks the enhancement request and links here in case anyone wants to see where this left off. Perhaps seeing that there is no PR open will encourage someone to take it up. Thanks all for your work so far!

**Pull request closed without merging**